File: sdk/python/src/digipay/webhook.py
"""HMAC signature verification for incoming DigiPay webhooks.

Framework-agnostic: hand it the raw bytes + the header + the secret,
get back a parsed event dict, or a :class:`DigiPayError` on mismatch.
"""

from __future__ import annotations

import hashlib
import hmac
import json

from .errors import DigiPayError
from .types import WebhookEvent


def verify_webhook(
    raw_body: bytes | str,
    signature: str | None,
    secret: str,
) -> WebhookEvent:
    """Verify the HMAC signature on an incoming DigiPay webhook and
    return the parsed event.

    Parameters
    ----------
    raw_body:
        The un-parsed bytes of the HTTP body, exactly as received. Do
        **not** re-serialize the JSON before passing it in — the HMAC
        covers the bytes, so any reserialization breaks the check. In
        Flask: ``request.get_data()``. In FastAPI: ``await request.body()``.
    signature:
        Value of the ``X-DigiPay-Signature`` header. Format is
        ``sha256=<hex>``; the prefix is tolerated if stripped.
    secret:
        The store's webhook secret. Treat like a password.

    Raises
    ------
    DigiPayError
        With ``status=401`` on missing / mismatched signature,
        ``status=400`` on malformed JSON after a valid signature.

    Example (Flask)
    ---------------
        from flask import Flask, request, abort
        from digipay import verify_webhook, DigiPayError

        app = Flask(__name__)

        @app.post("/digipay-webhook")
        def webhook():
            try:
                event = verify_webhook(
                    raw_body=request.get_data(),
                    signature=request.headers.get("X-DigiPay-Signature"),
                    secret=os.environ["DIGIPAY_SECRET"],
                )
            except DigiPayError as err:
                abort(err.status)
            # event["event"] == "session.paid" | "session.confirmed" | ...
            return "", 200
    """
    if not signature:
        raise DigiPayError("Missing X-DigiPay-Signature header", 401)
    if not secret:
        raise DigiPayError("Webhook secret is required", 400)

    # Header format is "sha256=<hex>"; tolerate a stripped prefix just
    # in case a proxy rewrote the header (uncommon but harmless here).
    provided = signature[len("sha256="):] if signature.startswith("sha256=") else signature

    body_bytes = raw_body.encode("utf-8") if isinstance(raw_body, str) else raw_body
    expected = hmac.new(secret.encode("utf-8"), body_bytes, hashlib.sha256).hexdigest()

    # hmac.compare_digest is constant-time and handles unequal lengths
    # safely — preferred over == for any secret comparison.
    if not hmac.compare_digest(expected, provided):
        raise DigiPayError("Webhook signature mismatch", 401)

    try:
        return json.loads(body_bytes.decode("utf-8"))
    except (json.JSONDecodeError, UnicodeDecodeError) as err:
        raise DigiPayError("Webhook body is not valid JSON", 400, err) from err

File: sdk/node/src/webhook.ts
import { createHmac, timingSafeEqual } from 'node:crypto';
import { DigiPayError } from './errors.js';
import type { WebhookEvent } from './types.js';

export interface VerifyWebhookOptions {
    /**
     * Raw request body, exactly as DigiPay sent it. Must be the un-parsed
     * bytes — every reserialization changes the signature. In Express,
     * mount `express.raw({ type: 'application/json' })` on the route.
     */
    rawBody: Buffer | string;
    /** Value of the `X-DigiPay-Signature` request header. */
    signature: string | undefined | null;
    /**
     * The webhook secret you got from `POST /v1/pay/merchants` (or rotated
     * via the dashboard). Treat it like a password.
     */
    secret: string;
}

/**
 * Verify the HMAC signature on an incoming DigiPay webhook and return the
 * parsed event. Throws `DigiPayError` (status 401) on missing / mismatched
 * signature so an Express handler can pass the error straight to the
 * default error pipeline.
 *
 *     app.post('/webhook',
 *       express.raw({ type: 'application/json' }),
 *       (req, res) => {
 *         try {
 *           const event = verifyWebhook({
 *             rawBody: req.body,
 *             signature: req.headers['x-digipay-signature'],
 *             secret: process.env.DIGIPAY_SECRET,
 *           });
 *           // event.event === 'session.paid' etc.
 *           res.status(200).end();
 *         } catch (err) {
 *           res.status(401).end();
 *         }
 *       });
 */
export function verifyWebhook(options: VerifyWebhookOptions): WebhookEvent {
    const { rawBody, signature, secret } = options;
    if (!signature) {
        throw new DigiPayError('Missing X-DigiPay-Signature header', 401);
    }
    if (!secret) {
        throw new DigiPayError('Webhook secret is required', 400);
    }

    // Header format: "sha256=<hex>". Tolerate the prefix being absent
    // since some proxies strip it.
    const provided = signature.startsWith('sha256=') ? signature.slice(7) : signature;

    const bodyBuf = Buffer.isBuffer(rawBody) ? rawBody : Buffer.from(rawBody, 'utf8');
    const expected = createHmac('sha256', secret).update(bodyBuf).digest('hex');

    // timingSafeEqual throws if the two buffers differ in length, which
    // would itself be a side-channel — guard with the length check first.
    const a = Buffer.from(expected);
    const b = Buffer.from(provided);
    if (a.length !== b.length || !timingSafeEqual(a, b)) {
        throw new DigiPayError('Webhook signature mismatch', 401);
    }

    try {
        return JSON.parse(bodyBuf.toString('utf8')) as WebhookEvent;
    } catch (err) {
        throw new DigiPayError('Webhook body is not valid JSON', 400, err);
    }
}
